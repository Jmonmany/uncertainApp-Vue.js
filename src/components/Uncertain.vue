<template>
  <img v-if="image" :src="image" alt="" />
  <div class="bg-dark"></div>
  <div class="uncertain-container">
    <input
      type="text"
      placeholder="Ask me a question"
      v-model="question"
      @keyup.enter="getQuestion()"
    />
    <p>Do not forget to write the question mark (?)</p>
    <div v-if="isValid">
      <h2>{{ question }}</h2>
      <h1>{{ answer }}</h1>
    </div>
  </div>
</template>

<script>
export default {
  props: {},
  name: "Uncertain",
  data() {
    return {
      question: "",
      answer: null,
      image: null,
      isValid: false,
    };
  },
  methods: {
    async getAnswer() {
      try {
        this.answer = "Waiting for answer...";
        const response = await fetch("https://yesno.wtf/api");
        const { answer, image } = await response.json();
        this.answer = answer;
        this.image = image;
      } catch (error) {
        this.answer = "Fail to fetch";
        this.img = null;
      }
    },
  },
  watch: {
    question(value, oldValue) {
      console.log({ value });
      if (!value.includes("?")) return;
      this.isValid = true;
      this.getAnswer();
    },
  },
};
</script>

<style scoped>
img,
.bg-dark {
  height: 100vh;
  left: 0px;
  max-height: 67%;
  max-width: 100%;
  position: fixed;
  bottom: 0px;
  width: 100vw;
}

.bg-dark {
  background-color: rgba(0, 0, 0, 0.4);
}

.uncertain-container {
  margin-block-start: 1rem;
  position: relative;
  z-index: 99;
}

input {
  width: 250px;
  padding: 10px 15px;
  border-radius: 5px;
  border: 1px solid black;
}
input:focus {
  outline: none;
}

p {
  color: white;
  font-size: 20px;
  margin-top: 0px;
}

h1,
h2 {
  color: white;
}

h2 {
  margin-top: 150px;
}
</style>
