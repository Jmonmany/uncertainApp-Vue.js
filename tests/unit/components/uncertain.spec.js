import Uncertain from "@/components/Uncertain.vue";
import { shallowMount, mount } from "@vue/test-utils";

describe("Uncertain component", () => {
  let wrapper;
  let clgSpy;
  global.fetch = jest.fn(() =>
    Promise.resolve({
      json: () => Promise.resolve({
        answer: "yes",
        forced: false,
        image: "https://yesno.wtf/assets/yes/1.gif"
      }),
    })
  );
  // mockear fetch
  beforeEach(() => {
    wrapper = shallowMount(Uncertain);
    clgSpy = jest.spyOn(console, "log");
    jest.clearAllMocks();
  });
  test("should match with snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
  test("should not fire anything when writing on input field", async () => {
    const getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer");
    const input = wrapper.find("input");
    await input.setValue("Hello");
    expect(clgSpy).toHaveBeenCalled();
    expect(getAnswerSpy).not.toHaveBeenCalled();
    // console.log(wrapper.vm);
  });
  test("should fire getAnswer as soon as writing ? mark", async () => {
    const getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer");
    const input = wrapper.find("input");
    await input.setValue("Hello?");
    expect(getAnswerSpy).toHaveBeenCalled();
  });
  test("getAnswer tests", async () => {
    console.log(wrapper.vm);
    await wrapper.vm.getAnswer();
    expect(wrapper.vm.answer).toBe("yes");
    expect(wrapper.vm.image).toBe("https://yesno.wtf/assets/yes/1.gif");
    // expect(wrapper.vm.forced).toBe(false); // no funciona
  });
  test("should trow error when fetching fails", async () => {
    fetch.mockImplementation(() => Promise.reject("API is down"))

    await wrapper.vm.getAnswer();

    const img = wrapper.find("img");
    const answer = wrapper.vm.answer
    expect(img.exists()).toBeFalsy();
    expect(answer).toBe("Fail to fetch");
  });
});
