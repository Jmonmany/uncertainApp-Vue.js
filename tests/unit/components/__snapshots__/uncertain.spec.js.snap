// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Uncertain component should match with snapshot 1`] = `
<!--v-if-->
<div class="bg-dark"></div>
<div class="uncertain-container"><input type="text" placeholder="Ask me a question">
  <p>Do not forget to write the question mark (?)</p>
  <!--v-if-->
</div>
`;
